week7/answers.sql

```diff
@@ -1,3 +1,309 @@
+	SELECT
+		c.character_id,
+		c.name AS character_name,	
+		i.name AS item_name,
+		i.armor,
+		i.damage
+	FROM characters c
+		INNER JOIN inventory inv
+			ON c.character_id = inv.character_id
+		INNER JOIN items i
+			ON inv.item_id = i.item_id
+	UNION
+	SELECT 
+		c.character_id,
+		c.name AS character_name,	
+		i.name AS item_name,
+		i.armor,
+		i.damage
+	FROM characters c
+		INNER JOIN equipped eq
+			ON c.character_id = eq.character_id
+		INNER JOIN items i
+			ON eq.item_id = i.item_id
+	ORDER BY character_id ASC, item_name ASC;
+
+
+CREATE OR REPLACE VIEW team_items AS
+	SELECT
+		t.team_id,
+		t.name AS team_name,
+		i.name AS item_name,
+		i.armor,
+		i.damage
+	FROM teams t
+		INNER JOIN team_members tm
+			ON t.team_id = tm.team_id
+		INNER JOIN inventory inv 
+			ON tm.character_id = inv.character_id
+		INNER JOIN items i
+			ON inv.item_id = i.item_id
+UNION 
+	SELECT
+		t.team_id,
+		t.name AS team_name,
+		i.name AS item_name,
+		i.armor,
+		i.damage
+	FROM teams t
+		INNER JOIN team_members tm
+			ON t.team_id = tm.team_id
+		INNER JOIN equipped eq
+			ON tm.character_id = eq.character_id
+		INNER JOIN items i
+			ON eq.item_id = i.item_id
+		ORDER BY team_id, item_name;
+
+
+DELIMITER ;;
+
+CREATE FUNCTION armor_total(char_id INT UNSIGNED)
+RETURNS INT UNSIGNED
+DETERMINISTIC
+BEGIN
+	DECLARE armor_points_equipped INT UNSIGNED;
+    DECLARE armor_points_cs INT UNSIGNED;
+    
+    SELECT SUM(i.armor) INTO armor_points_equipped
+		FROM characters c
+			INNER JOIN character_stats cs
+				ON c.character_id = cs.character_id
+			INNER JOIN equipped eq
+				ON c.character_id = eq.character_id
+			INNER JOIN items i
+				ON eq.item_id = i.item_id
+	WHERE c.character_id = char_id;
+    
+	SELECT cs.armor INTO armor_points_cs
+		FROM characters c
+			INNER JOIN character_stats cs
+				ON c.character_id = cs.character_id
+	WHERE c.character_id = char_id;
+        
+	return armor_points_equipped + armor_points_cs;
+END;;
+DELIMITER ;
+
+DELIMITER ;;
+CREATE PROCEDURE attack(IN char_attacked_id INT UNSIGNED, IN id_item_equipped INT UNSIGNED)
+BEGIN
+    	DECLARE char_armor INT SIGNED;
+    DECLARE char_health INT;
+	DECLARE attack_damage INT SIGNED;
+    DECLARE difference INT;
+    DECLARE new_char_health INT;
+    
+	-- character armor 
+    SELECT armor_total(char_attacked_id) INTO char_armor;
+    
+    -- character_health
+    SELECT health INTO char_health FROM character_stats WHERE character_id = char_attacked_id;
+    
+    
+    -- item damage and attack damage - character armor
+    SELECT i.damage INTO attack_damage
+        FROM items i
+	    INNER JOIN equipped eq
+	        ON eq.item_id = i.item_id
+    WHERE eq.equipped_id = id_item_equipped;
+
+    SET difference = attack_damage - char_armor;
+    SET new_char_health = char_health - difference;
+    
+    -- check what happens with item damage
+    IF difference > 0 THEN
+	    UPDATE character_stats SET health=new_char_health WHERE character_id = char_attacked_id;
```

Comment:
This is where you are getting an error in your checks. In the `charater_stats` table, you define the `health` column as being `UNSIGNED`, however on this line the value of `new_char_health` can be negative (which cannot be represented by an `UNSIGNED` character). I recommend checking whether `new_char_health` is greater than `0` (then UPDATE stats) or if it is `0` or less (then `DELETE` character). You have all of the parts here, but you just need to restructure your `IF` statements. Let me know if you need further assistance.